### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by AlyAtalla**
General requirements:
there are no linter errors;
correct flow Gitflow is used;
work is documented in a professional way.

Project requirements:
We installed and set up CanCanCan in project:

added it in Gemfile
ran bundle install
ran rails generate cancan:ability to generate abilites.rb
updated abilites.rb file to meet our app logic
We added a role column to the users table:

rails g migration AddRoleToUsers role:string
ran rails db:migrate to generate new schema.rb
We added delete action to PostsController and CommentsController.

We added the "Delete" button to the views and ensured only authorised users could see it.
We also ensured only authorised users can see "Create New Post", "Comment" and "Like" buttons.

Result:
A user who is not logged in can:

only read the contents of the website
A logged-in user can:

create new posts on their page
like any posts
comment on any posts
delete a post if it is theirs or if they have an admin role (column role has value "admin")
delete a comment if it is theirs or if they have an admin role (column role has value "admin")

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -2,21 +2,44 @@ class Ability
...
-    end
+    # if the user is logged in can create posts, comments and likes
+    can :create, Post
+    can :create, Comment
+    can :create, Like
+
+    can :destroy, Post, author_id: user.id # if the user is logged in can delete it's own posts
+    can :destroy, Comment, user_id: user.id # if the user is logged in can delete it's own comments
+    can :destroy, Like, user_id: user.id # if the user is logged in can delete it's own likes
+    return unless user.role == 'admin'
```
Comment (codepantha) [inline app/models/ability.rb]: - [ ] [OPTIONAL] Please consider creating a `user.admin?` method inside your `user` model as this makes your code cleaner and you can call it anywhere you want to ensure that a user is an admin 👍

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!



## Highlights 🎯 
✔️ Descriptive PR
✔️ Added CanCanCan Gem in Gemfile 📌 
✔️ Added `role` column in the users table 📌 
✔️ No linter errors 🟢 

## Suggested changes
_Check the comments under the review._

- [ ] The `bio` and `photo` fields are preventing new account creation (sign ups). As a result, some of the features of your authorization project cannot be tested. Please fix this bug 🐛 
image

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/models/ability.rb [hunk 2]
```diff
@@ -2,21 +2,44 @@ class Ability
...
+
+    can :manage, :all # give all remaining permissions only to the admins
```
Comment (Reem-lab) [inline app/models/ability.rb]: - [ ] Great job working on the functionality so far but I think after signing up the first user and the second user. so the second user can delete the post for the first user and that shouldn't be the case the second user just should add like and comments on the posts for the first user and he can just delete his own post and his own comments ✨🚀
so here you can add a condition 

         if user.admin? 
         can :manage, :all
         else
so great you added  ` can :manage, :all` but you should include it inside that condition so not all can manage these postsو comments for the other users just the admin✨✅

**Review by Reem-lab — CHANGES_REQUESTED** [review]
# Hi @AlyAtalla, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet. ✅✨🚀


 ## Highlights

- [ ] No Linters error 🚀
- [ ]  Followed GitFlow 💯
- [ ]  Great for adding the ability class functionality👏🏻
- [ ]  Descriptive PR 💯

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by ridwanediallo — APPROVED** [review]
# STATUS: APPROVED 🟢
## Hi @AlyAtalla , 
Congratulations!GIF

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used GitHub flow correctly. ✔
- Well done implementing the requested changes. 🎉 🎉 🎉 ✔

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**